src/Type/Zend/ServiceManagerGetDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace ZendPhpStan\Type\Zend;
+
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\NullType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+use Zend\ServiceManager\ServiceManager;
+
+final class ServiceManagerGetDynamicReturnTypeExtension implements \PHPStan\Type\DynamicMethodReturnTypeExtension
+{
+    /**
+     * @var ServiceManagerLoader
+     */
+    private $serviceManagerLoader;
+
+    public function __construct(ServiceManagerLoader $serviceManagerLoader)
+    {
+        $this->serviceManagerLoader = $serviceManagerLoader;
+    }
+
+    public function getClass(): string
+    {
+        return ServiceManager::class;
+    }
+
+    public function isMethodSupported(MethodReflection $methodReflection): bool
+    {
+        return 'get' === $methodReflection->getName();
+    }
+
+    public function getTypeFromMethodCall(
+        MethodReflection $methodReflection,
+        MethodCall $methodCall,
+        Scope $scope
+    ): Type {
+        if (1 !== \count($methodCall->args)) {
+            return new MixedType();
+        }
+
+        $argType = $scope->getType($methodCall->args[0]->value);
+        if (! $argType instanceof ConstantStringType) {
+            return new MixedType();
+        }
+
+        $objectName     = $argType->getValue();
+        $serviceManager = $this->serviceManagerLoader->getServiceManager();
+        if (! $serviceManager->has($objectName)) {
+            return new NullType();
```

Comment:
@ondrejmirtes this resambles [Doctrine/GetRepositoryDynamicReturnTypeExtension:58](https://github.com/phpstan/phpstan-doctrine/blob/master/src/Type/Doctrine/GetRepositoryDynamicReturnTypeExtension.php#L58): if I don't check the existence of the service and I directly try to get it, like phpunit-doctrine does, I get a not so user friendly error like:

```
 ------ --------------------------------------------------------------------------------------------------------------------- 
  Line   serviceManagerDynamicReturn.php                                                                                      
 ------ --------------------------------------------------------------------------------------------------------------------- 
         Internal error: Unable to resolve service "bar" to a factory; are you certain you provided it during configuration?  
         Run PHPStan with --debug option and post the stack trace to:                                                         
         https://github.com/phpstan/phpstan/issues/new                                                                        
 ------ --------------------------------------------------------------------------------------------------------------------- 
```
For the code https://github.com/Slamdunk/phpstan-zend-framework/pull/1/files#diff-850327538197c025d0724fadcc005fbbR26

I would like to customize the error message, and most important, get the line that generate the error.
Is this feasible? The final result I would like to achieve is:
```
 ------ ------------------------------------------------------------------------------------------------------
  Line   serviceManagerDynamicReturn.php                                                                      
 ------ ------------------------------------------------------------------------------------------------------
    26   Unable to resolve service "bar" to a factory; are you certain you provided it during configuration?  
 ------ ------------------------------------------------------------------------------------------------------
```

Reply:
You need not to throw an exception from the extension, and you need to write a rule that checks for this situation.

Reply:
Thank you :+1: